### ISSUE TITLE
Kernel error scanning with journalctl

### ISSUE THREAD

**Issue by fredoh9** [thesofproject/sof-test#292]
Marc suggested to switch to journalctl while reviewing https://github.com/thesofproject/sof-test/pull/290

Current we grep for negative keywords, if it hit any then return error. These are current keywords for error.
```
err_str="error|failed|timed out|panic|oops"

dmesg | grep -E "$err_str" | grep -vE "$ignore_str"
```
Later we should drop such "screen scraping" and use something more robust like `journalctl --dmesg --priority err` with no false positive (for instance a debug message with one of these keywords) and no false negative (e.g. ERR:)

Links:
- https://gitlab.devtools.intel.com/sof-ci/sofcitest/-/issues/500 Make the complete output of "journalctl --quiet --boot 0 -p warning" easily available in each test result

**Comment by fredoh9** [comment]
PR is already available.
https://github.com/thesofproject/sof-test/pull/262

**Comment by xiulipan** [comment]
I think this is already done by https://github.com/thesofproject/sof-test/pull/529 @fredoh9 Please confirm and close.

**Comment by fredoh9** [comment]
yes, thank you!

**Comment by marc-hb** [comment]
I'm re-opening this because we're still using `dmesg` to capture kernel logs in test reports, example: https://github.com/thesofproject/linux/issues/3731#issuecomment-1170821163

`dmesg` has number of issues:
- it's a ring buffer so it may be incomplete
- `ktime` timestamps are relatively useless because difficult to relate to user space
- no way to combine with user space logs if/when desired

**Comment by ujfalusi** [comment]
@marc-hb, what about `dmesg --level=err` ?

**Comment by marc-hb** [comment]
> what is wrong with dmesg --level=err to filter the prints?

Everything I listed above? :-)

> We still need to filter SOF erros (likely) 

Please prove me wrong but so far I haven't found any `dmesg` feature that `journalctl` does not have (and of course either can be piped to `grep` the same)

**Comment by ujfalusi** [comment]
@marc-hb, sure, `journalctl` is an OS while `dmesg` is a command line tool from util-linux.
> it's a ring buffer so it may be incomplete

and journalctl takes the kernel log from the same ring buffer...

> It's not available after a reboot

journalctl saves the logs to somewhere, so any syslog tools (/var/log/kernel.log.* in my machines)

> ktime timestamps are relatively useless because difficult to relate to user space

`dmesg --time-format ctime`
I think the issue is that the timestamps are hard to relate tot he firmware stamps.

> It's missing the hostname

Hrm true, but I usually know where I have run it.

> no way to combine with user space logs if/when desired

`dmesg -u`

If you think `journalctl` will give us better CI reports, I'm all for it, don't get me wrong.
But running the tests on my own machines are getting harder and harder.
I'll remain:
```
$ journalctl
zsh: command not found: journalctl
```

**Comment by marc-hb** [comment]
> and journalctl takes the kernel log from the same ring buffer...

I don't understand sorry. I just ran both `dmesg` and `journal -k` one test device that has been running some sof-test with verbose sof logs:

```
root@up2:~# dmesg | wc -l
2908
root@up2:~# journalctl -k | wc -l
14600
```

I bet you'll observe the same thing with `/var/log/whatever`

> dmesg --time-format ctime

I just tried this and I found a 72 seconds gap with `journalctl -k`, any idea why? Suspend/resume?

After a LOT of time spent troubleshooting NTP issues, I have 100% confidence in `journalctl` timestamps, of course as long as NTP is working https://github.com/thesofproject/sof-test/discussions/908)


>  Hrm true, but I usually know where I have run it.

"works4me" ? :-)

> dmesg -u

This is new to me. I tried it and the output was very surprising: it seems to have only `systemd` logs, no other user space logs and almost no kernel logs.

> If you think journalctl will give us better CI reports, I'm all for it, don't get me wrong.

Yes for the reasons listed above.

sof-test has _already_ a very strong dependency on `journalctl` (see old comments above) so there's no reason to stop short and keep using `dmesg` in just one sof-test place bar all the others.

> zsh: command not found: journalctl

I'm sorry but we have barely bandwidth to validate only one distribution right now (without even pulseaudio / pipewire...), a growing backlog and like it or not  `journalctl` is definitely the standard. As long as your logging solution has an equivalent to  journalctl's `--since` option then it should be possible to implement some abstraction layer, please send patches! This will also help with:
- #684 
- https://chromium.googlesource.com/chromiumos/docs/+/HEAD/logging.md (rsyslogd now?)

> I think the issue is that the timestamps are hard to relate to the firmware stamps.

That's a very good point. The fw timestamp is typically initialized from `ktime` but I suspect it can then diverge depending on D3 , innacurate frequency guess in sof-logger and what not? If you find a way to have BOTH ktime and wall clock in captured logs then I'm all for it but as long as we can't easily relate to FW timestamps then it's better to match _at least_ user space logs.

BTW what is the simplest way to read `ktime` from a shell? I tried a Python one-liner in 46dadd0c563e6c but I'm afraid it's not reliable :-(

**Comment by marc-hb** [comment]
BTW I recently found that Ubuntu 22 still uses `rsyslogd` in _addition_ to journald. Weird.

- https://github.com/thesofproject/sof-test/discussions/908#discussioncomment-3809638
